File: data_structure/tree/src/TreeDemo.js
// JavaScript
// データ構造: 木 (Tree)

class NodeData {
    constructor(value) {
        this._value = value;
        this._parent = null;
        this._children = [];
    }
    
    get_value() {
        return this._value;
    }

    get_parent() {
        return this._parent;
    }

    get_children() {
        return this._children;
    }

    set_parent(parent) {
        this._parent = parent;
        return true;
    }

    add_child(child) {
        child.set_parent(this);
        this._children.push(child);
        return true;
    }
    
    remove_child(child) {
        const index = this._children.indexOf(child);
        if (index !== -1) {
            child.set_parent(null);
            this._children.splice(index, 1);
            return true;
        } else {
            return false;
        }
    }
    
    is_leaf() {
        return this._children.length === 0;
    }
}

class TreeData {
    constructor() {
        this._data = null;
    }

    get() {
        return this._data;
    }
    
    get_height(node = null) {
        if (node === null) {
            node = this._data;
        }
        if (node === null) {
            return 0;
        }
        if (node.is_leaf()) {
            return 1;
        }
        return 1 + Math.max(...node.get_children().map(child => this.get_height(child)));
    }
    
    get_parent(node) {
        return node.get_parent();
    }
    
    get_children(node) {
        return node.get_children();
    }

    get_node(value, node = null) {
        if (node === null) {
            node = this._data;
        }
        if (node === null) {
            return null;
        }
        
        if (node.get_value() === value) {
            return node;
        }
        
        for (const child of node.get_children()) {
            const result = this.get_node(value, child);
            if (result !== null) {
                return result;
            }
        }
        
        return null;
    }

    add(parent, value) {
        const new_node = new NodeData(value);
        if (parent === null) {
            if (this._data === null) {
                this._data = new_node;
                return true;
            } else {
                console.log(`ERROR: ${value} 重複です`);
                return false;
            }
        } else {
            parent.add_child(new_node);
            return true;
        }
    }

    remove(node) {
        if (node === null) {
            return false;
        }
        
        if (node === this._data) {
            this._data = null;
            return true;
        }
        
        const parent = node.get_parent();
        if (parent !== null) {
            return parent.remove_child(node);
        }
        
        return false;
    }
    
    traverse(node = null, mode = "pre-order") {
        if (node === null) {
            node = this._data;
        }
        if (node === null) {
            return [];
        }
        
        const result = [];
        
        if (mode === "pre-order") {
            result.push(node.get_value());
            for (const child of node.get_children()) {
                result.push(...this.traverse(child, mode));
            }
        } else if (mode === "post-order") {
            for (const child of node.get_children()) {
                result.push(...this.traverse(child, mode));
            }
            result.push(node.get_value());
        } else if (mode === "level-order") {
            const queue = [node];
            while (queue.length > 0) {
                const current = queue.shift();
                result.push(current.get_value());
                queue.push(...current.get_children());
            }
        }
        
        return result;
    }
    
    is_leaf(node) {
        return node !== null && node.is_leaf();
    }
    
    is_empty() {
        return this._data === null;
    }
    
    size(node = null) {
        if (node === null) {
            node = this._data;
        }
        if (node === null) {
            return 0;
        }
        
        let count = 1;
        for (const child of node.get_children()) {
            count += this.size(child);
        }
        
        return count;
    }
    
    clear() {
        this._data = null;
        return true;
    }

    display() {
        if (this._data === null) {
            return [];
        }
        
        return this.traverse(null, "level-order");
    }
}

function main() {
    console.log("Tree TEST -----> start");

    console.log("\nnew");
    const tree_data = new TreeData();
    console.log(`  現在のデータ: ${tree_data.display()}`);

    console.log("\nis_empty");
    const output1 = tree_data.is_empty();
    console.log(`  出力値: ${output1}`);

    console.log("\nsize");
    const output2 = tree_data.size();
    console.log(`  出力値: ${output2}`);

    console.log("\nadd");
    const input_params1 = [null, 'Root'];
    console.log(`  入力値: ${input_params1}`);
    const output3 = tree_data.add(...input_params1);
    console.log(`  出力値: ${output3}`);
    console.log(`  現在のデータ: ${tree_data.display()}`);
    
    const root_node = tree_data.get();

    console.log("\nadd");
    const input_params2 = [root_node, 'Child1'];
    console.log(`  入力値: ${input_params2}`);
    const output4 = tree_data.add(...input_params2);
    console.log(`  出力値: ${output4}`);
    console.log(`  現在のデータ: ${tree_data.display()}`);

    console.log("\nadd");
    const input_params3 = [root_node, 'Child2'];
    console.log(`  入力値: ${input_params3}`);
    const output5 = tree_data.add(...input_params3);
    console.log(`  出力値: ${output5}`);
    console.log(`  現在のデータ: ${tree_data.display()}`);

    console.log("\nget_node");
    const input_value = 'Child1';
    console.log(`  入力値: ${input_value}`);
    const output6 = tree_data.get_node(input_value);
    console.log(`  出力値: ${output6}`);
    console.log(`  現在のデータ: ${tree_data.display()}`);

    console.log("\ntraverse");
    const input_mode = 'pre-order';
    console.log(`  入力値: ${input_mode}`);
    const output7 = tree_data.traverse(null, input_mode);
    console.log(`  出力値: ${output7}`);
    console.log(`  現在のデータ: ${tree_data.display()}`);

    console.log("\nTree TEST <----- end");
}

main();

File: data_structure/tree/src/TreeDemo.swift
// Swift
// データ構造: 木 (Tree)

class NodeData {
    private var _value: String
    private var _parent: NodeData?
    private var _children: [NodeData] = []
    
    init(_ value: String) {
        self._value = value
    }
    
    func getValue() -> String {
        return _value
    }
    
    func getParent() -> NodeData? {
        return _parent
    }
    
    func getChildren() -> [NodeData] {
        return _children
    }
    
    func setParent(_ parent: NodeData?) -> Bool {
        self._parent = parent
        return true
    }
    
    func addChild(_ child: NodeData) -> Bool {
        child.setParent(self)
        _children.append(child)
        return true
    }
    
    func removeChild(_ child: NodeData) -> Bool {
        if let index = _children.firstIndex(where: { $0 === child }) {
            child.setParent(nil)
            _children.remove(at: index)
            return true
        } else {
            return false
        }
    }
    
    func isLeaf() -> Bool {
        return _children.isEmpty
    }
}

class TreeData {
    private var _data: NodeData?
    
    func get() -> NodeData? {
        return _data
    }
    
    func getHeight(node: NodeData? = nil) -> Int {
        let currentNode = node ?? _data
        guard let currentNode = currentNode else {
            return 0
        }
        
        if currentNode.isLeaf() {
            return 1
        }
        
        return 1 + currentNode.getChildren().map { getHeight(node: $0) }.max() ?? 0
    }
    
    func getParent(_ node: NodeData) -> NodeData? {
        return node.getParent()
    }
    
    func getChildren(_ node: NodeData) -> [NodeData] {
        return node.getChildren()
    }
    
    func getNode(_ value: String, node: NodeData? = nil) -> NodeData? {
        let currentNode = node ?? _data
        guard let currentNode = currentNode else {
            return nil
        }
        
        if currentNode.getValue() == value {
            return currentNode
        }
        
        for child in currentNode.getChildren() {
            if let result = getNode(value, node: child) {
                return result
            }
        }
        
        return nil
    }
    
    func add(_ parent: NodeData?, _ value: String) -> Bool {
        let newNode = NodeData(value)
        if parent == nil {
            if _data == nil {
                _data = newNode
                return true
            } else {
                print("ERROR: \(value) 重複です")
                return false
            }
        } else {
            parent!.addChild(newNode)
            return true
        }
    }
    
    func remove(_ node: NodeData?) -> Bool {
        guard let node = node else {
            return false
        }
        
        if node === _data {
            _data = nil
            return true
        }
        
        if let parent = node.getParent() {
            return parent.removeChild(node)
        }
        
        return false
    }
    
    func traverse(node: NodeData? = nil, mode: String = "pre-order") -> [String] {
        let currentNode = node ?? _data
        guard let currentNode = currentNode else {
            return []
        }
        
        var result: [String] = []
        
        switch mode {
        case "pre-order":
            result.append(currentNode.getValue())
            for child in currentNode.getChildren() {
                result.append(contentsOf: traverse(node: child, mode: mode))
            }
        case "post-order":
            for child in currentNode.getChildren() {
                result.append(contentsOf: traverse(node: child, mode: mode))
            }
            result.append(currentNode.getValue())
        case "level-order":
            var queue: [NodeData] = [currentNode]
            while !queue.isEmpty {
                let current = queue.removeFirst()
                result.append(current.getValue())
                queue.append(contentsOf: current.getChildren())
            }
        default:
            break
        }
        
        return result
    }
    
    func isLeaf(_ node: NodeData?) -> Bool {
        return node != nil && node!.isLeaf()
    }
    
    func isEmpty() -> Bool {
        return _data == nil
    }
    
    func size(node: NodeData? = nil) -> Int {
        let currentNode = node ?? _data
        guard let currentNode = currentNode else {
            return 0
        }
        
        var count = 1
        for child in currentNode.getChildren() {
            count += size(node: child)
        }
        
        return count
    }
    
    func clear() -> Bool {
        _data = nil
        return true
    }
    
    func display() -> [String] {
        if _data == nil {
            return []
        }
        
        return traverse(mode: "level-order")
    }
}

func main() {
    print("Tree TEST -----> start")
    
    print("\nnew")
    let treeData = TreeData()
    print("  現在のデータ: \(treeData.display())")
    
    print("\nis_empty")
    let isEmpty = treeData.isEmpty()
    print("  出力値: \(isEmpty)")
    
    print("\nsize")
    let size = treeData.size()
    print("  出力値: \(size)")
    
    print("\nadd")
    let inputParams1 = (nil, "Root")
    print("  入力値: \(inputParams1)")
    let output1 = treeData.add(inputParams1.0, inputParams1.1)
    print("  出力値: \(output1)")
    print("  現在のデータ: \(treeData.display())")
    
    let rootNode = treeData.get()
    
    print("\nadd")
    let inputParams2 = (rootNode, "Child1")
    print("  入力値: \(inputParams2)")
    let output2 = treeData.add(inputParams2.0, inputParams2.1)
    print("  出力値: \(output2)")
    print("  現在のデータ: \(treeData.display())")
    
    print("\nadd")
    let inputParams3 = (rootNode, "Child2")
    print("  入力値: \(inputParams3)")
    let output3 = treeData.add(inputParams3.0, inputParams3.1)
    print("  出力値: \(output3)")
    print("  現在のデータ: \(treeData.display())")
    
    print("\nget_node")
    let inputValue = "Child1"
    print("  入力値: \(inputValue)")
    let output4 = treeData.getNode(inputValue)
    print("  出力値: \(output4 != nil ? "some data" : "nil")")
    print("  現在のデータ: \(treeData.display())")
    
    print("\ntraverse")
    let inputMode = "pre-order"
    print("  入力値: \(inputMode)")
    let output5 = treeData.traverse(mode: inputMode)
    print("  出力値: \(output5)")
    print("  現在のデータ: \(treeData.display())")
    
    print("\nTree TEST <----- end")
}

main()